soc: nordic: nrf52/nrf53: use DT to configure regulators

gmarull opened the pull request:
Check commit messages

gmarull committed:
dts: bindings: regulator: add nordic,nrf5[2]x-regulator[-hv]

Some Nordic SoCs, like nRF52 contain an internal regulator for the main
SoC supply. Depending on the SoC, the regulator can have multiple
configurations (e.g. single/double stage), which mainly depends on
supporting "high-voltage" mode or not (VDDH pin supply). This patch adds
bindings for nRF5X regulator and nRF52X HV regulator.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: nordic: nrf52x: instantiate regulators

The SoC main supply is part of the POWER IP block. The POWER IP is a
kind of multi-purpose block, so each of its functions is described as a
child node in DT, like similar other MFD. This allows to have specific
properties, e.g. for DC/DC mode.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf52: add support for DT-based regulators config

In addition to Kconfig options (soon to be deprecated), allow
configuring the regulators using DT.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf52*: use DT to configure regulators

Instead of board level options selecting SOC_DCDC_NRF52X or
SOC_DCDC_NRF52X_HV.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: usb: use DT regulator properties

Instead of CONFIG_SOC_DCDC_NRF52X[_HV], about to be deprecated.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf52: deprecate SOC_DCDC_NRF52X[_HV]

Main supply can now be configured using DT.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

nordicjm commented on dts/bindings/regulator/nordic,nrf-main-supply-hv.yaml:
```diff
@@ -0,0 +1,23 @@
+
+include: nordic,nrf-main-supply-common.yaml
+
+properties:
+  nordic,hv-enable:
+    type: boolean
+    description: |
+      Enable high-voltage mode, i.e. supply from VDDH pin. Note that this
+      option is not available on SoCs with QFN48 package, where VDDH is
+      internally shorted to VDD.
```
there should be a voltage selection for this too

nordicjm replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/70088/files#diff-e2b7be8df6625aa6ab076a44777f8cc8c323ec90a1c2394d7c3168a945744b5e

gmarull replied:
this is set via UICR, so not regulator registers.

gmarull replied:
actually I likely prefer our DT representation, let me give it a spin

fabiobaltieri replied:
are you going to put it under the custom binding? I also think it does not fit very well with the regulator api since what you can do with it is very limited (can't turn it on an off, need to reset to change the voltage...) but it'd be cool to have it as a property here and drop the board code

gmarull replied:
```
dcdcen: dcdcen@40000578 {
	compatible = "regulator-fixed";
	regulator-name = "dcdcen";
	reg = <0x40000578 0x1>;
};

dcdcen0: dcdcen0@40000580 {
	compatible = "nordic,hvdcdc-regulator";
	regulator-name = "dcdcen0";
	reg = <0x40000580 0x1>;
	regulator-min-microvolt = <1800000>;
	regulator-max-microvolt = <3300000>;
};
```

is not a bad description, however, the `regulator-min/max-microvolt` properties are wrong, as they are meant to limit the operating range at runtime, something first unknown at soc level and then, something you can't do in practice (it's a UICR setting). As far as I can see, REGOUT0 is not currenly handled either. Another problem is that users would be dealing with dcdcen nodelabels, but these are registers in the end. Regulators are called REG/REG0/REG1. What about:

```
regulators {
	REG0: regulator@40000580 {
		compatible = "nordic,nrf-hv-regulator";
		reg = <0x40000580 0x1>;
		regulator-name = "REG0";
		status = "disabled";
	};

	REG1: regulator@40000578 {
		compatible = "regulator-fixed";
		reg = <0x40000578 0x1>;
		regulator-name = "REG1";
		regulator-init-mode = <NRF_REG_LDO>;
	};
};
```

then, at board level:

```
/* to set DC/DC instead of LDO (default) */
&REG1 {
	regulator-init-mode = <NRF_REG_DCDC>;
};

/* to enable HV mode */
&REG0 {
	status = "okay";
};
```

in the future, we could also enable support for:

```
&REG0 {
    regulator-init-microvolt = <1800000>;
};
```

thoughts?

nordicjm replied:
Looks good

gmarull committed:
dts: bindings: regulator: add nordic,nrf53x-regulators

nRF53X regulator IP is specific to that series, eg, not equal to nRF91X.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: regulator: add nordic,nrf53x-regulator-hv

nRF53X HV regulator differs from eg nRF52X as it offers a silent mode
option. For this reason, a new compatible is used, even if now such
capability is not exposed yet.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: nordic: nrf5340: instantiate regulators

Instantiate all available regulators: VREGMAIN, VREGRADIO and VREGH.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf53: allow configuring regulators using DT

Instead of Kconfig options which are about to be deprecated.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf53*: use DT to configure regulators

Instead of board level options selecting SOC_DCDC_NRF53X*.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf53: deprecate SOC_DCDC_NRF53X*

Regulators can now be configured using DT.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: regulator: add nordic,nrf91x-regulators

To describe nRF91X specific REGULATORS IP.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
doc: releases: migration-guide: add nRF52/53 regulator changes

Inform users about Kconfig options deprecation and show what is the
alternative with some examples.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

fabiobaltieri reviewed (commented):
Looks good, will obviously not go in 3.7, you may want to start a file for the 4.0 migration guide

gmarull commented:
> Looks good, will obviously not go in 3.7, you may want to start a file for the 4.0 migration guide

will do after release, this PR will likely conflict with the network core boot anyway.

aasinclair commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -83,6 +83,47 @@ Serial
+
+      /* enable REG0 (HV mode) */
+      &reg0 {
+          satus = "okay";
```
```suggestion
          status = "okay";
```

aasinclair commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -83,6 +83,47 @@ Serial
+
+      /* enable VREGH (HV mode) */
+      &vregh {
+          satus = "okay";
```
```suggestion
          status = "okay";
```

aasinclair reviewed (changes requested):
A couple of doc typos, but otherwise looks very nice.

rerickson1 reviewed (approved):
Ezurio boards look good.
